coordinatorlib/src/main/scala/com/socrata/datacoordinator/truth/DatabaseAccessors.scala

```diff
@@ -435,18 +439,40 @@ object DatasetMutator {
         }
       }
 
-      def validateRollup(ru: RollupInfo): Unit = {
-        val prefixedSchema: Seq[(ColumnName, CT)] =
-          datasetMap.schema(ru.copyInfo).values.map { colInfo =>
-            val prefix = if (colInfo.userColumnId.underlying.startsWith(":")) "" else "_"
-            (new ColumnName(prefix + colInfo.userColumnId.underlying), colInfo.typ)
-          }(collection.breakOut)
+      private def getPrefixedDsContext(copyInfo: CopyInfo): com.socrata.soql.environment.DatasetContext[CT] = {
+        val prefixedSchema: Seq[(ColumnName, CT)] = datasetMap.schema(copyInfo).values.map { colInfo =>
+          val prefix = if (colInfo.userColumnId.underlying.startsWith(":")) "" else "_"
+          (new ColumnName(prefix + colInfo.userColumnId.underlying), colInfo.typ)
+        }(collection.breakOut)
         val prefixedDsContext = new com.socrata.soql.environment.DatasetContext[CT] {
           val schema: OrderedMap[ColumnName, CT] = OrderedMap(prefixedSchema: _*)
         }
+        prefixedDsContext
+      }
+
+      def validateRollup(ru: RollupInfo): Unit = {
+        val prefixedDsContext = getPrefixedDsContext(ru.copyInfo)
         val analyzer = soqlAnalyzer
         try {
-          analyzer.analyzeFullQuery(ru.soql)(Map(TableName.PrimaryTable.qualifier -> prefixedDsContext))
+          val parsed = new Parser(AbstractParser.defaultParameters).binaryTreeSelect(ru.soql)
+          val tableNames = collectTableNames(parsed)
+          val initContext = Map(TableName.PrimaryTable.qualifier -> prefixedDsContext)
+          val contexts = tableNames.foldLeft(initContext) { (acc, tn) =>
+            val tableName = TableName(tn)
+            val ctx = for {
+              dsInfo <- datasetMapReader.datasetInfoByResourceName(ResourceName(tableName.nameWithSodaFountainPrefix))
+              copyInfo <- datasetMap.published(dsInfo)
+            } yield {
+              getPrefixedDsContext(copyInfo)
+            }
+            ctx match {
+              case Some(context) =>
+                acc + (tn -> context)
+              case None =>
+                acc
+            }
+          }
+          analyzer.analyzeBinary(parsed)(contexts)
```

Comment:
Ok, so, checking my understanding:
* we need to pass all the various tables (not just the primary table) in context, so that rollups can work cross-table
* in order to extract all the table names, we need to work with the query in binaryTree format instead of soql format

* but in the end, we're still calling analyze with the context to validate
yes?

Reply:
That's right